src/ralph/util/models.py

```diff
@@ -109,3 +108,194 @@ def save(self, mute=False, visited=None, *args, **kwargs):
+    value = db.IntegerField()
+
+
+class CustomAttributeStringValue(CustomAttributeValue):
+    """String value for custom attribute"""
+    value = db.CharField(max_length=255)
+
+
+class CustomAttributeOption(db.Model):
+    """An option for choice fields."""
+    custom_attribute = db.ForeignKey(CustomAttribute)
+    value = db.CharField(max_length=255)
+
+
+class CustomAttributeSingleChoiceValue(CustomAttributeValue):
+    """Single choice value for custom attribute"""
+    value = db.ForeignKey(CustomAttributeOption)
+
+
+CA_TYPE2MODEL = {
+    CustomAttributeTypes.INTEGER.id: CustomAttributeIntegerValue,
+    CustomAttributeTypes.STRING.id: CustomAttributeStringValue,
+    CustomAttributeTypes.SINGLE_CHOICE.id: CustomAttributeSingleChoiceValue,
+}
+
+
+class WithCustomAttributesMeta(type(db.Model)):
+    """Metaclass for models with custom attributes"""
+
+    def __init__(cls, clsname, bases, dict_):
+        super(WithCustomAttributesMeta, cls).__init__(clsname, bases, dict_)
+        cls._custom_attributes = None
+
+    @property
+    def content_type(cls):
+        return ContentType.objects.get_for_model(cls)
+
+    @property
+    def custom_attributes(cls):
+        if cls._custom_attributes is None:
+            cls.refresh_custom_attributes()
+        return cls._custom_attributes
+
+    def refresh_custom_attributes(cls):
+        if cls._custom_attributes is not None:
+            for ca in cls._custom_attributes:
+                delattr(cls, ca.name)
+        new_custom_attributes = CustomAttribute.objects.filter(
+            content_type=cls.content_type
+        )
+        cls._custom_attributes = new_custom_attributes
+        for ca in new_custom_attributes:
+            setattr(cls, ca.name, ca)
+        cls.attrs_loaded = True
+
+
+class WithCustomAttributes(db.Model):
+    """Mixin for classes with custom attributes"""
+
+    __metaclass__ = WithCustomAttributesMeta
+
+    class Meta:
+        abstract = True
+
+    def __init__(self, *args, **kwargs):
+        self.ca_values = None
+        super(WithCustomAttributes, self).__init__(*args, **kwargs)
+
+    def load_custom_attributes(self):
+        """Loads the values for custom attributes into the __dict__ of this
+        model. This method will be called once on first attempt to access the
+        custom attribute."""
+        self.ca_values = {}
+        for type_id, Model in CA_TYPE2MODEL.items():
+            values_set = Model.objects.filter(object_id=self.pk)
+            self.ca_values[type_id] = {}
+            for value in values_set:
+                value.dirty = False
+                value.to_delete = False
+                self.ca_values[type_id][value.attribute.pk] = value
+        for ca in type(self).custom_attributes:
+            value = self.ca_values[ca.type].get(ca.pk)
+            if value is not None:
+                self.__dict__[ca.name] = value.value
+            else:
+                self.__dict__[ca.name] = None
```

Comment:
I really really don't like this idea, as it would exploit the fact that getattr is broken.